### PR TITLE
Graph Services sequence diagram

### PR DESCRIPTION
**Pull request by jachro**
closes https://github.com/SwissDataScienceCenter/renku-graph/issues/39

### CODE REVIEW

File: docs/_static/uml/graph-push-event-sequence.uml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    == Push Event ==
+    Client->Webhook: POST /webhooks/events
+    activate Webhook
+    Webhook->Tokens: GET /projects/:id/tokens
+    activate Tokens
+    Tokens->Webhook: OK with Access Token or NOT_FOUND
+    deactivate Tokens
+    group Not processed commits finding process
+    Webhook->Webhook ++ : find commits ids not in the Event Log
+    Webhook->Log: find the youngest event
```
Comment (rokroskar) [inline docs/_static/uml/graph-push-event-sequence.uml]:
```suggestion
    Webhook->Log: find the most recent event
```

File: docs/_static/uml/graph-push-event-sequence.uml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    activate Log
+    Log->Webhook: EventId of the youngest event or nothing
```
Comment (rokroskar) [inline docs/_static/uml/graph-push-event-sequence.uml]:
```suggestion
    Log->Webhook: EventId of the most recent event or nothing
```

File: docs/_static/uml/graph-push-event-sequence.uml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    deactivate Log
+    Webhook->GitLab: GET /api/v4/projects/:id/repository/commits/:id
+    activate GitLab
+    GitLab->Webhook: OK with commit info
+    deactivate GitLab
+    Webhook->Log: filter out parent commits which are already in the Event Log yet
```
Comment (rokroskar) [inline docs/_static/uml/graph-push-event-sequence.uml]:
```suggestion
    Webhook->Log: filter out parent commits which are already in the Event Log
```

File: docs/_static/uml/graph-push-event-sequence.uml [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    activate Log
+    Log->Webhook: Parent commitIds not in the Event Log
+    deactivate Log
+    return Commit Events not in the Event Log yet
```
Comment (rokroskar) [inline docs/_static/uml/graph-push-event-sequence.uml]:
```suggestion
    return Commit Events not found in the Event Log
```

File: docs/developer/services/graph_service.rst [hunk 5]
```diff
@@ -17,3 +17,30 @@ architecture is illustrated below.
...
+
+Sequence diagram of Graph Services APIs and processes.
+""""""""""""""""""""""""""""""""""""""""""""""""""""""
+POST <webhook-service>/projects/:id/webhooks/validation
+
+.. uml:: ../../_static/uml/graph-validate-hook-sequence.uml
+
+POST <webhook-service>/projects/:id/webhooks
+
+.. uml:: ../../_static/uml/graph-create-hook-sequence.uml
```
Comment (rokroskar) [inline docs/developer/services/graph_service.rst]: I think putting "create webhook" before "validate webhook" would make sense

File: docs/developer/services/graph_service.rst [hunk 6]
```diff
@@ -17,3 +17,30 @@ architecture is illustrated below.
...
+
+GET <webhook-service>/projects/:id/events/status
+
+.. uml:: ../../_static/uml/graph-events-status-sequence.uml
+
+POST <webhook-service>/webhooks/events
```
Comment (rokroskar) [inline docs/developer/services/graph_service.rst]: the "events" endpoint should come before "status"

File: docs/_static/uml/graph-create-hook-sequence.uml [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    actor Client
+    participant "Webhook\nService" as Webhook
+    participant GitLab
+    participant "Tokens\nRepository" as Tokens
+    database "Event\nLog" as Log
+
+    == Webhook creation - hook does exists ==
+    Client->Webhook: POST /projects/:id/webhooks
+    activate Webhook
+    ref over Webhook, Log: Hook Validation as presented above returning hook exists
```
Comment (rokroskar) [inline docs/_static/uml/graph-create-hook-sequence.uml]:
```suggestion
    ref over Webhook, Log: Hook Validation as presented above returning OK
```

File: docs/_static/uml/graph-create-hook-sequence.uml [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    Webhook->Client: OK
+    deactivate Webhook
+
+    == Webhook creation - hook does not exist ==
+    Client->Webhook: POST /projects/:id/webhooks
+    activate Webhook
+    ref over Webhook, Log: Hook Validation as presented above returning hook does not exist
```
Comment (rokroskar) [inline docs/_static/uml/graph-create-hook-sequence.uml]:
```suggestion
    ref over Webhook, Log: Hook Validation as presented above returning NOT_FOUND
```

File: docs/_static/uml/graph-create-hook-sequence.uml [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    deactivate GitLab
+    Webhook->GitLab: POST /api/v4/projects/:id/hooks
+    activate GitLab
+    GitLab->Webhook: OK
+    deactivate GitLab
+    Webhook->Tokens: PUT /projects/:id/tokens
+    activate Tokens
+    Tokens->Webhook: NO_CONTENT
+    deactivate Tokens
+    group Events history loading
```
Comment (rokroskar) [inline docs/_static/uml/graph-create-hook-sequence.uml]: does this mean that the hook creation triggers event processing?

**Comment by jachro** [comment]
I addressed all the comments and in terms of the status codes I somehow find it more appropriate to use wording instead of "magic numbers" :) What I've just done is I put codes next to the names in e.g. OK (200).

**Pull request merged**